### PR TITLE
fix: pypika translators [TCTC-3109]

### PR DESCRIPTION
**Pull request by Sanix-Darker**
## PYPIKA STEPS SCAN

This is the list of "broken", "working" or not "implemented" steps (depending on fixtures) for postgres-pypika and redshift !

#### POSTGRESQL-PYPIKA

<details><summary>GET THE LIST</summary>
##### WORKING & FIXED

- [x] tests/backends/fixtures/select/simple.json
- [x] tests/backends/fixtures/aggregate/simple.json
- [x] tests/backends/fixtures/aggregate/simple_aggregate_with_null.json
- [x] tests/backends/fixtures/comparetext/simple.json
- [x] tests/backends/fixtures/delete/non_existing_column.json
- [x] tests/backends/fixtures/todate/automatic_guess.json
- [x] tests/backends/fixtures/todate/from_timestamp.json
- [x] tests/backends/fixtures/top/with_groups.json
- [x] tests/backends/fixtures/filter/dates.json
- [x] tests/backends/fixtures/customsql/simple.json
- [x] tests/backends/fixtures/convert/to_text.json
- [x] tests/backends/fixtures/convert/postgres_redshift_to_float.json
- [x] tests/backends/fixtures/convert/postgres_redshift_to_integer.json
- [x] tests/backends/fixtures/formula/simple.json
- [x] tests/backends/fixtures/formula/division.json
- [x] tests/backends/fixtures/formula/nested.json

---

###### CONVERT
- [ ] tests/backends/fixtures/convert/to_bool.json
- [ ] tests/backends/fixtures/convert/to_datetime_from_timestamp.json
- [ ] tests/backends/fixtures/convert/to_datetime.json
- [ ] tests/backends/fixtures/convert/to_integer_from_datetime.json
- [ ] tests/backends/fixtures/convert/to_integer_only_floats.json
- [ ] tests/backends/fixtures/convert/to_integer_only_strings.json

###### AGGREGATE
- [ ] tests/backends/fixtures/aggregate/first.json
- [ ] tests/backends/fixtures/aggregate/last.json
- [ ] tests/backends/fixtures/aggregate/keep_original_granularity_empty_on.json
- [ ] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations.json
- [ ] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations_multiple_columns.json
- [ ] tests/backends/fixtures/aggregate/count_with_null.json

###### FILTER
- [ ] tests/backends/fixtures/filter/dates_from_until.json
- [ ] tests/backends/fixtures/filter/dates_from_until_strip_hours.json
- [ ] tests/backends/fixtures/filter/relative_dates.json

---

###### FROM_DATE
- [ ] tests/backends/fixtures/fromdate/simple.json
- [ ] tests/backends/fixtures/fromdate/day_longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_space.json

###### FORMULA
- [ ] tests/backends/fixtures/formula/division_zero.json
- [ ] tests/backends/fixtures/formula/nested_divisions.json
- [ ] tests/backends/fixtures/formula/special_col.json

###### IFTHENELSE
- [ ] tests/backends/fixtures/ifthenelse/dates-conditions.json
- [ ] tests/backends/fixtures/ifthenelse/regex.json

###### TO_DATE
- [ ] tests/backends/fixtures/todate/simple.json
- [ ] tests/backends/fixtures/todate/int_as_year_with_automatic_guess.json
- [ ] tests/backends/fixtures/todate/str_as_year_with_automatic_guess.json

###### CUMSUM (started an implementation, but more tweaking needed)
- [ ] tests/backends/fixtures/cumsum/simple.json
- [ ] tests/backends/fixtures/cumsum/with_groups.json
- [ ] tests/backends/fixtures/cumsum/with_multiple_values.json

---

###### MOVINGAVERAGE
- [ ] tests/backends/fixtures/movingaverage/simple.json
- [ ] tests/backends/fixtures/movingaverage/with_groups.json

###### PERCENTAGE
- [ ] tests/backends/fixtures/percentage/simple.json
- [ ] tests/backends/fixtures/percentage/advanced.json

###### PIVOT
- [ ] tests/backends/fixtures/pivot/simple.json
- [ ] tests/backends/fixtures/pivot/columns_cleaning.json

###### RANK
- [ ] tests/backends/fixtures/rank/simple.json
- [ ] tests/backends/fixtures/rank/group_by.json

###### ROLLUP
- [ ] tests/backends/fixtures/rollup/simple.json
- [ ] tests/backends/fixtures/rollup/groupby.json
- [ ] tests/backends/fixtures/rollup/no_aggregation.json

---

###### TOTALS
- [ ] tests/backends/fixtures/totals/simple.json

###### WATERFALL
- [ ] tests/backends/fixtures/waterfall/simple.json
- [ ] tests/backends/fixtures/waterfall/with_aggregation.json
- [ ] tests/backends/fixtures/waterfall/with_groups.json

###### DATE_EXTRACT
- [ ] tests/backends/fixtures/dateextract/after_add_missing_dates_duplicate_values.json
- [ ] tests/backends/fixtures/dateextract/all_cases.json
- [ ] tests/backends/fixtures/dateextract/legacy_config.json
- [ ] tests/backends/fixtures/dateextract/legacy_config_without_column.json
- [ ] tests/backends/fixtures/dateextract/mongo_all_cases.json

###### EVOLUTION
- [ ] tests/backends/fixtures/evolution/abs_nogroups.json
- [ ] tests/backends/fixtures/evolution/perc_groups.json
- [ ] tests/backends/fixtures/evolution/perc_nogroups.json

###### STATISTCIS
- [ ] tests/backends/fixtures/statistics/all.json
---
</details>

#### REDSHIFT

<details><summary>GET THE LIST</summary>
##### WORKING & FIXED

- [x] tests/backends/fixtures/select/simple.json
- [x] tests/backends/fixtures/delete/non_existing_column.json
- [x] tests/backends/fixtures/todate/automatic_guess.json
- [x] tests/backends/fixtures/todate/from_timestamp.json
- [x] tests/backends/fixtures/top/with_groups.json
- [x] tests/backends/fixtures/aggregate/simple_aggregate_with_null.json
- [x] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations.json
- [x] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations_multiple_columns.json
- [x] tests/backends/fixtures/comparetext/simple.json
- [x] tests/backends/fixtures/customsql/simple.json
- [x] tests/backends/fixtures/filter/dates.json
- [x] tests/backends/fixtures/formula/simple.json
- [x] tests/backends/fixtures/formula/division.json
- [x] tests/backends/fixtures/formula/nested.json
- [x] tests/backends/fixtures/convert/to_text.json
- [x] tests/backends/fixtures/convert/redshift_to_bool.json
- [x] tests/backends/fixtures/convert/postgres_redshift_to_float.json
- [x] tests/backends/fixtures/convert/postgres_redshift_to_integer.json
---

###### CONVERT
- [ ] tests/backends/fixtures/convert/to_datetime.json
- [ ] tests/backends/fixtures/convert/to_datetime_from_timestamp.json
- [ ] tests/backends/fixtures/convert/to_integer_from_datetime.json
- [ ] tests/backends/fixtures/convert/to_integer_only_floats.json
- [ ] tests/backends/fixtures/convert/to_integer_only_strings.json

###### AGGREGATE
- [ ] tests/backends/fixtures/aggregate/simple.json
- [ ] tests/backends/fixtures/aggregate/first.json
- [ ] tests/backends/fixtures/aggregate/keep_original_granularity_empty_on.json
- [ ] tests/backends/fixtures/aggregate/last.json
- [ ] tests/backends/fixtures/aggregate/count_with_null.json

###### FROM_DATE
- [ ] tests/backends/fixtures/fromdate/simple.json

###### TO_DATE
- [ ] tests/backends/fixtures/todate/simple.json
---

###### FILTER
- [ ] tests/backends/fixtures/filter/dates_from_until.json
- [ ] tests/backends/fixtures/filter/dates_from_until_strip_hours.json
- [ ] tests/backends/fixtures/filter/relative_dates.json

###### FORMULA
- [ ] tests/backends/fixtures/formula/division_zero.json
- [ ] tests/backends/fixtures/formula/nested_divisions.json
- [ ] tests/backends/fixtures/formula/special_col.json

###### FROM_DATE
- [ ] tests/backends/fixtures/fromdate/day_longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_space.json

###### IFTHENELSE
- [ ] tests/backends/fixtures/ifthenelse/dates-conditions.json
- [ ] tests/backends/fixtures/ifthenelse/regex.json
---

###### TO_DATE
- [ ] tests/backends/fixtures/todate/int_as_year_with_automatic_guess.json
- [ ] tests/backends/fixtures/todate/str_as_year_with_automatic_guess.json

###### CUMSUM
- [ ] tests/backends/fixtures/cumsum/simple.json
- [ ] tests/backends/fixtures/cumsum/with_groups.json
- [ ] tests/backends/fixtures/cumsum/with_multiple_values.json
---

###### MOVINGAVERAGE
- [ ] tests/backends/fixtures/movingaverage/simple.json
- [ ] tests/backends/fixtures/movingaverage/with_groups.json

###### PERCENTAGE
- [ ] tests/backends/fixtures/percentage/simple.json
- [ ] tests/backends/fixtures/percentage/advanced.json

###### PIVOT
- [ ] tests/backends/fixtures/pivot/simple.json
- [ ] tests/backends/fixtures/pivot/columns_cleaning.json

###### RANK
- [ ] tests/backends/fixtures/rank/simple.json
- [ ] tests/backends/fixtures/rank/group_by.json
---
###### ROLLUP
- [ ] tests/backends/fixtures/rollup/simple.json
- [ ] tests/backends/fixtures/rollup/groupby.json
- [ ] tests/backends/fixtures/rollup/no_aggregation.json

###### TOTALS
- [ ] tests/backends/fixtures/totals/simple.json

###### WATERFALL
- [ ] tests/backends/fixtures/waterfall/simple.json
- [ ] tests/backends/fixtures/waterfall/with_aggregation.json
- [ ] tests/backends/fixtures/waterfall/with_groups.json

###### DATE_EXTRACT
- [ ] tests/backends/fixtures/dateextract/after_add_missing_dates_duplicate_values.json
- [ ] tests/backends/fixtures/dateextract/all_cases.json
- [ ] tests/backends/fixtures/dateextract/legacy_config.json
- [ ] tests/backends/fixtures/dateextract/legacy_config_without_column.json
- [ ] tests/backends/fixtures/dateextract/mongo_all_cases.json
---
###### EVOLUTION
- [ ] tests/backends/fixtures/evolution/abs_nogroups.json
- [ ] tests/backends/fixtures/evolution/perc_groups.json
- [ ] tests/backends/fixtures/evolution/perc_nogroups.json

###### STATISTCIS
- [ ] tests/backends/fixtures/statistics/all.json
---
</details>

#### ATHENA

Some fixtures have been implemented so far for athena here : 

<details><summary>GET THE LIST</summary>
#### IMPLEMENTED AND WORKING

(Most of these step where made with a specific dataset, that's why we needed here to have a suffix ("athena"))

- [x] tests/backends/fixtures/ifthenelse/combinations_athena.json
- [x] tests/backends/fixtures/ifthenelse/nested_athena.json
- [x] tests/backends/fixtures/ifthenelse/simple_athena.json
- [x] tests/backends/fixtures/aggregate/count_athena.json
- [x] tests/backends/fixtures/aggregate/count_distinct_athena.json
- [x] tests/backends/fixtures/aggregate/with_original_granularity_athena.json
- [x] tests/backends/fixtures/aggregate/without_on_athena.json
- [x] tests/backends/fixtures/formula/division_athena.json
- [x] tests/backends/fixtures/formula/nested_athena.json
- [x] tests/backends/fixtures/formula/simple_athena.json
- [x] tests/backends/fixtures/rename/edge_case_athena.json
- [x] tests/backends/fixtures/rename/multiple_athena.json
- [x] tests/backends/fixtures/rename/simple_athena.json
- [x] tests/backends/fixtures/filter/eq_athena.json
- [x] tests/backends/fixtures/trim/multiple_columns_athena.json
- [x] tests/backends/fixtures/argmax/simple_athena.json
- [x] tests/backends/fixtures/argmax/with_group_athena.json
- [x] tests/backends/fixtures/argmin/simple_athena.json
- [x] tests/backends/fixtures/delete/simple_athena.json
- [x] tests/backends/fixtures/duplicate/simple_athena.json
- [x] tests/backends/fixtures/fillna/simple_athena.json
- [x] tests/backends/fixtures/lowercase/simple_athena.json
---
- [x] tests/backends/fixtures/replace/simple_athena.json
- [x] tests/backends/fixtures/sort/simple_athena.json
- [x] tests/backends/fixtures/split/simple_athena.json
- [x] tests/backends/fixtures/substring/simple_athena.json
- [x] tests/backends/fixtures/text/simple_athena.json
- [x] tests/backends/fixtures/top/simple_athena.json
- [x] tests/backends/fixtures/uniquegroups/simple_athena.json
- [x] tests/backends/fixtures/uppercase/simple_athena.json
- [x] tests/backends/fixtures/trim/single_column_athena.json
---

###### AGGREGATE
- [ ] tests/backends/fixtures/aggregate/count_with_null.json
- [ ] tests/backends/fixtures/aggregate/first.json
- [ ] tests/backends/fixtures/aggregate/keep_original_granularity_empty_on.json
- [ ] tests/backends/fixtures/aggregate/last.json
- [ ] tests/backends/fixtures/aggregate/legacy_syntax.json
- [ ] tests/backends/fixtures/aggregate/simple_aggregate_with_null.json
- [ ] tests/backends/fixtures/aggregate/simple.json
- [ ] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations.json
- [ ] tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations_multiple_columns.json
---
###### COMPARETEXT
- [ ] tests/backends/fixtures/comparetext/simple.json

###### CONCATENATE
- [ ] tests/backends/fixtures/concatenate/simple.json

###### CONVERT
- [ ] tests/backends/fixtures/convert/to_bool.json
- [ ] tests/backends/fixtures/convert/to_datetime_from_timestamp.json
- [ ] tests/backends/fixtures/convert/to_datetime.json
- [ ] tests/backends/fixtures/convert/to_float.json
- [ ] tests/backends/fixtures/convert/to_integer_from_datetime.json
- [ ] tests/backends/fixtures/convert/to_integer.json
- [ ] tests/backends/fixtures/convert/to_integer_only_floats.json
- [ ] tests/backends/fixtures/convert/to_integer_only_strings.json
- [ ] tests/backends/fixtures/convert/to_text.json

###### CUMSUM
- [ ] tests/backends/fixtures/cumsum/simple.json
- [ ] tests/backends/fixtures/cumsum/with_groups.json
- [ ] tests/backends/fixtures/cumsum/with_multiple_values.json
---
###### CUSTOM
- [ ] tests/backends/fixtures/custom/simple.json

###### CUSTOMSQL
- [ ] tests/backends/fixtures/customsql/simple.json

###### DATE_EXTRACT
- [ ] tests/backends/fixtures/dateextract/after_add_missing_dates_duplicate_values.json
- [ ] tests/backends/fixtures/dateextract/all_cases.json
- [ ] tests/backends/fixtures/dateextract/legacy_config.json
- [ ] tests/backends/fixtures/dateextract/legacy_config_without_column.json
- [ ] tests/backends/fixtures/dateextract/mongo_all_cases.json

###### DISOLVE
- [ ] tests/backends/fixtures/dissolve/pandas_advanced.json
- [ ] tests/backends/fixtures/dissolve/pandas_no_aggs.json
- [ ] tests/backends/fixtures/dissolve/pandas_simple.json

###### DURATION
- [ ] tests/backends/fixtures/duration/simple.json
---
###### EVOLUTION
- [ ] tests/backends/fixtures/evolution/abs_nogroups.json
- [ ] tests/backends/fixtures/evolution/perc_groups.json
- [ ] tests/backends/fixtures/evolution/perc_nogroups.json

###### FILLNA
- [ ] tests/backends/fixtures/fillna/date.json
- [ ] tests/backends/fixtures/fillna/fillinteger.json

###### FILTER
- [ ] tests/backends/fixtures/filter/dates_from_until.json
- [ ] tests/backends/fixtures/filter/dates_from_until_strip_hours.json
- [ ] tests/backends/fixtures/filter/dates.json
- [ ] tests/backends/fixtures/filter/eq.json
- [ ] tests/backends/fixtures/filter/relative_dates.json
---
###### FORMULA
- [ ] tests/backends/fixtures/formula/division_zero.json
- [ ] tests/backends/fixtures/formula/nested_divisions.json
- [ ] tests/backends/fixtures/formula/special_col.json

###### FROM_DATE
- [ ] tests/backends/fixtures/fromdate/day_longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/day_shortmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/longmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_dash.json
- [ ] tests/backends/fixtures/fromdate/shortmonth_year_space.json
- [ ] tests/backends/fixtures/fromdate/simple.json

###### IFTHENELSE
- [ ] tests/backends/fixtures/ifthenelse/dates_conditions.json
- [ ] tests/backends/fixtures/ifthenelse/regex.json
---
###### MOVINGAVERAGE
- [ ] tests/backends/fixtures/movingaverage/simple.json
- [ ] tests/backends/fixtures/movingaverage/with_groups.json

###### PERCENTAGE
- [ ] tests/backends/fixtures/percentage/advanced.json
- [ ] tests/backends/fixtures/percentage/simple.json

###### PIVOT
- [ ] tests/backends/fixtures/pivot/columns_cleaning.json
- [ ] tests/backends/fixtures/pivot/simple.json

###### RANK
- [ ] tests/backends/fixtures/rank/group_by.json
- [ ] tests/backends/fixtures/rank/simple.json
---
###### ROLLUP
- [ ] tests/backends/fixtures/rollup/groupby.json
- [ ] tests/backends/fixtures/rollup/no_aggregation.json
- [ ] tests/backends/fixtures/rollup/simple.json

###### STATISTCIS
- [ ] tests/backends/fixtures/statistics/all.json

###### TO_DATE
- [ ] tests/backends/fixtures/todate/automatic_guess.json
- [ ] tests/backends/fixtures/todate/from_timestamp.json
- [ ] tests/backends/fixtures/todate/int_as_year_with_automatic_guess.json
- [ ] tests/backends/fixtures/todate/simple.json
- [ ] tests/backends/fixtures/todate/str_as_year_with_automatic_guess.json

###### TOP
- [ ] tests/backends/fixtures/top/with_groups.json
---
###### TOTALS
- [ ] tests/backends/fixtures/totals/simple.json
- [ ] tests/backends/fixtures/totals/simple-pandas.json

###### UNPIVOT
- [ ] tests/backends/fixtures/unpivot/pandas.json

###### WATERFALL
- [ ] tests/backends/fixtures/waterfall/simple.json
- [ ] tests/backends/fixtures/waterfall/with_aggregation.json
- [ ] tests/backends/fixtures/waterfall/with_groups.json
---
</details>

### STEPS SKIPPED (for now, they are not implemented)
<details><summary>GET THE LIST</summary>
##### APPEND
- [ ] tests/backends/fixtures/append/domain_name_only.json
- [ ] tests/backends/fixtures/append/with_domain_name.json
- [ ] tests/backends/fixtures/append/simple.json

##### JOIN
- [ ] tests/backends/fixtures/join/domain_name_only.json
- [ ] tests/backends/fixtures/join/inner_geo_left.json
- [ ] tests/backends/fixtures/join/inner_geo_right.json
- [ ] tests/backends/fixtures/join/inner.json
- [ ] tests/backends/fixtures/join/left_geo_left.json
- [ ] tests/backends/fixtures/join/left_geo_right.json
- [ ] tests/backends/fixtures/join/left.json
- [ ] tests/backends/fixtures/join/left_outer.json
</details>

### CODE REVIEW

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 1]
```diff
@@ -146,6 +152,14 @@ def _get_aggregate_function(
         self: Self, agg_function: "AggregateFn"
     ) -> functions.AggregateFunction:
         match agg_function:
+            # Commenting this since postgres and redshift
```
Comment (raphaelvignes) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:155]: By doing so, you forbid the use of these functions for translators where it's implemented no? 
Perhaps, you could update the match case based on the instance's Dialect.

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 2]
```diff
@@ -305,6 +315,38 @@ def get_sql(self, **kwargs: Any) -> str:
         # without actually executing the query
         return custom_query, StepTable(columns=["*"])
 
+    def cumsum(
```
Comment (raphaelvignes) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:318]: Instead of having a broken step for Snowflake and all others, perhapse you could: 
1. Override the method for Snowflake Translator
2. Raise unimplemented for others.

**Review by raphaelvignes — CHANGES_REQUESTED** [review]
Thanks for this work!
I see some minor changes to do.

**Review by lukapeschke — CHANGES_REQUESTED** [review]
Thanks for the work! It's nice to make our tests more consistent!

Two general remarks:
* Since you've updated redshift's integration tests to run on the same dataset as athena and BQ, `redshift_pypika` should not be excluded from `_pypika` fixtures anymore.
* OTOH all fixtures that aren't `_pypika` should now exclude `redshift_pypika`. I found the following:
```
ag --json -L '"redshift_pypika"' server/tests/backends/fixtures/
server/tests/backends/fixtures/aggregate/without_on.json
server/tests/backends/fixtures/aggregate/count.json
server/tests/backends/fixtures/aggregate/legacy_syntax.json
server/tests/backends/fixtures/aggregate/simple_aggregate_with_null.json
server/tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations.json
server/tests/backends/fixtures/aggregate/with_original_granularity.json
server/tests/backends/fixtures/argmax/simple.json
server/tests/backends/fixtures/aggregate/with_original_granularity_multiple_aggregations_multiple_columns.json
server/tests/backends/fixtures/argmin/with_group.json
server/tests/backends/fixtures/comparetext/simple.json
server/tests/backends/fixtures/argmax/with_group.json
server/tests/backends/fixtures/argmin/simple.json
server/tests/backends/fixtures/convert/to_bool.json
server/tests/backends/fixtures/convert/to_text.json
server/tests/backends/fixtures/delete/simple.json
server/tests/backends/fixtures/duration/simple.json
server/tests/backends/fixtures/duplicate/simple.json
server/tests/backends/fixtures/fillna/date.json
server/tests/backends/fixtures/fillna/fillinteger.json
server/tests/backends/fixtures/fillna/simple.json
server/tests/backends/fixtures/filter/dates.json
server/tests/backends/fixtures/filter/eq.json
server/tests/backends/fixtures/ifthenelse/combinations.json
server/tests/backends/fixtures/ifthenelse/nested.json
server/tests/backends/fixtures/ifthenelse/simple.json
server/tests/backends/fixtures/lowercase/simple.json
server/tests/backends/fixtures/rename/edge_case.json
server/tests/backends/fixtures/aggregate/count_distinct.json
server/tests/backends/fixtures/rename/multiple.json
server/tests/backends/fixtures/rename/simple.json
server/tests/backends/fixtures/replace/simple.json
server/tests/backends/fixtures/sort/simple.json
server/tests/backends/fixtures/todate/from_timestamp.json
server/tests/backends/fixtures/split/simple.json
server/tests/backends/fixtures/text/simple.json
server/tests/backends/fixtures/todate/automatic_guess.json
server/tests/backends/fixtures/top/simple.json
server/tests/backends/fixtures/substring/simple.json
server/tests/backends/fixtures/trim/multiple_columns.json
server/tests/backends/fixtures/trim/single_column.json
server/tests/backends/fixtures/top/with_groups.json
server/tests/backends/fixtures/uniquegroups/simple.json
server/tests/backends/fixtures/uppercase/simple.json
server/tests/backends/fixtures/customsql/simple.json
server/tests/backends/fixtures/formula/division.json
server/tests/backends/fixtures/formula/nested.json
server/tests/backends/fixtures/formula/simple.json
server/tests/backends/fixtures/select/simple.json
server/tests/backends/fixtures/absolutevalue/simple.json
```

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 1 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py:155]: After checks, it was not in the list as you can see here : https://github.com/ToucanToco/weaverbird/blob/master/server/src/weaverbird/backends/pypika_translator/translators/base.py#L148
I just added this comment so that will keep in mind pypika has those built-in function and for next card adapt depending on the translator what works and what doesn't !

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 2 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: i will rather comment this function for now, since, it's not working on snowflake's edge cases, thanks to pointing this out 🙏🏾

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 1 above)
Comment (raphaelvignes) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py:154]: :+1:

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 3]
```diff
@@ -466,6 +507,16 @@ def formula(
...
         #   CAST("my age" AS float) + 1 / 2
+
+        # We add quotes to identify columns
+        # for c in table.columns:
+        #     step.formula = (
+        #         f" {step.formula} ".replace(f" {c} ", f" \"{c}\" ")
+        #         .replace(f"({c} ", f"(\"{c}\" ")
+        #         .replace(f" {c})", f" \"{c}\")")
+        #     )
+        #     step.formula = " ".join(step.formula.split())
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:518]: To me, this does not sound like something we want to do. Escaping column names should be up to PyPika, especially since the quoting character to use depends on the SQL dialect

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 4]
```diff
@@ -305,6 +314,38 @@ def get_sql(self, **kwargs: Any) -> str:
         # without actually executing the query
         return custom_query, StepTable(columns=["*"])
 
+    # def cumsum(
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:317]: If this cannot be implemented as is, should this comment really live here ? I'd rather have it in the snowflake translator.

File: server/src/weaverbird/backends/pypika_translator/translators/postgresql.py [hunk 5]
```diff
@@ -45,5 +47,24 @@ def duration(
...
+    def customsql(
+        self: Self, *, step: "CustomSqlStep", table: StepTable
+    ) -> tuple["QueryBuilder", StepTable]:
+        """create a custom sql step based on the current table named ##PREVIOUS_STEP## in the query"""
+
+        if table.name is None:
+            raise MissingTableNameError()
+
+        # We want all columns wraps in semicolumns
+        table.columns = [f'"{c}"' for c in table.columns]
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/postgresql.py:59]: This should be up to PyPika

File: server/tests/backends/sql_translator/common.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            s = df[colname]
+            idx = s.first_valid_index()
+            if idx is not None and (first_value := s.loc[idx]) is not None:
+                if isinstance(first_value, (Decimal, float)) and "." in str(first_value):
+                    _clean_float(first_value, colname)
+
+                if convert_nan_to_none:
+                    for i, v in enumerate(df[colname].values):
+                        if str(df[colname].values[i]).lower() == 'nan':
+                            df[colname].values[i] = None
```
Comment (lukapeschke) [inline server/tests/backends/sql_translator/common.py:42]: These functions seem dangerous to me, as we're applying some transformation to data that won't be applied when a pipeline is actually executed. We can keep `standardized_columns` and the  `_clean_float` part of `standardized_values` for now until postgres pypika tests are ISO with the rest, but I'm against the nan to None part for the following reasons:

1. The behaviour is undefined, depending on the pandas version: Either the series' type will change from `float` to object, or the `None` will be converted to `nan` again
2. `assert_dataframes_equal` should treat all null-ish values as equal (basically if `pd.isnull(v1) and pd.isnull(v2)` we can consider `v1 == v2`)

File: server/tests/backends/sql_translator_integration_tests/test_sql_redshift_translator_steps.py [hunk 7]
```diff
@@ -1,162 +1,76 @@
...
+_BEERS_TABLE_COLUMNS = [
+    "price_per_l",
+    "alcohol_degree",
+    "name",
+    "cost",
+    "beer_kind",
+    "volume_ml",
+    "brewing_date",
+    "nullable_name",
+]
```
Comment (lukapeschke) [inline server/tests/backends/sql_translator_integration_tests/test_sql_redshift_translator_steps.py:49]: Have you been able to insert the dataset into redshift ?

**Review by lukapeschke — CHANGES_REQUESTED** [review]

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 3 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: Yeah, exactly, that's why it's commented, i removed these lines (with comments) here : https://github.com/ToucanToco/weaverbird/pull/1390/commits/fda3338f1dc14a8243e54ffbd76caf3769329d11

File: server/src/weaverbird/backends/pypika_translator/translators/postgresql.py (hunk 5 above)
Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/postgresql.py]: True, i handled that here : https://github.com/ToucanToco/weaverbird/pull/1390/commits/570db8145897639ba5e0328258cb13cf6ccd969f

File: server/tests/backends/sql_translator/common.py (hunk 6 above)
Comment (Sanix-Darker) [reply server/tests/backends/sql_translator/common.py]: Alright, you got a point !
I removed that here : https://github.com/ToucanToco/weaverbird/pull/1390/commits/cdd2c0e860c23fae385fd1fbc0ffe70ea3588adf

But i had to skip the convert_text step since it was failing : 

```
>   ???
E   AssertionError: DataFrame.iloc[:, 0] (column name="value") are different
E
E   DataFrame.iloc[:, 0] (column name="value") values are different (16.66667 %)
E   [index]: [0, 1, 2, 3, 4, 5]
E   [left]:  [41, 42, 43.5, 43.6, nan, meh]
E   [right]: [41, 42, 43.5, 43.6, None, meh]

pandas/_libs/testing.pyx:167: AssertionError
================================================== short test summary info ===================================================
FAILED tests/backends/sql_translator_integration_tests/test_sql_postgres_pypika_translator_steps.py::test_sql_translator_pipeline[/convert_to_text]
```

File: server/tests/backends/sql_translator_integration_tests/test_sql_redshift_translator_steps.py (hunk 7 above)
Comment (Sanix-Darker) [reply server/tests/backends/sql_translator_integration_tests/test_sql_redshift_translator_steps.py:49]: not yet, it's skipped it's a subject of a card i created here : https://toucantoco.atlassian.net/jira/software/c/projects/TCTC/boards/18/backlog?view=detail&selectedIssue=TCTC-3204&issueLimit=100

File: server/tests/backends/sql_translator/common.py (hunk 6 above)
Comment (lukapeschke) [reply server/tests/backends/sql_translator/common.py]: Seems legit. We'd have to agree on a convention between all translators to decide what the value for null should be as text. Should it be an emtpy string, the litteral `"null"` something else... ? But that's out of the scope of this PR

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 4 above)
Comment (lukapeschke) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: great thx

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 8]
```diff
@@ -301,6 +310,11 @@ def get_sql(self, **kwargs: Any) -> str:
             query=step.query.replace("##PREVIOUS_STEP##", table.name),
         )
 
+        custom_query.query = (
+            custom_query.query.replace(";", "")
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:314]: why is this needed ?

Comment (Sanix-Darker) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: a missed !
I was to removed the ";" for custom queries, since, none-pypika fixture fixture was failing on it : https://github.com/ToucanToco/weaverbird/blob/master/server/tests/backends/fixtures/customsql/simple.json#L17

I just removed that : https://github.com/ToucanToco/weaverbird/pull/1390/commits/ea7945089bb792e44353dbf227340a2d770526a3

**Review by lukapeschke — APPROVED** [review]
:+1:

**Review by raphaelvignes — APPROVED** [review]
OK for me. 
Thx!
